src/twisted/mail/smtp.py

```diff
@@ -533,7 +542,8 @@ def do_MAIL(self, rest):
             return
         # Clear old recipient list
         self._to = []
-        m = self.mail_re.match(rest)
+
+        m = self.mail_re.match(rest.decode("utf-8"))
```

Comment:
This doesn't comply with RFC 6531.  [A client may send UTF-8 encoded sequences if and only if an ESMTP server advertises the `SMTPUTF8` extension](https://tools.ietf.org/html/rfc6531#section-3.2);

> This document applies when an SMTPUTF8-aware SMTP client or server
> supports the SMTPUTF8 extension.  **For all other cases**, and for
> addresses and messages that do not require an SMTPUTF8 extension,
> **SMTPUTF8-aware SMTP clients and servers do not change the behavior
> specified in RFC 5321** [[RFC5321](https://tools.ietf.org/html/rfc5321)].

Where RFC 5321 obsoletes RFC 2821, the RFC called out in the previous implementation.

In this PR, the `SMTP` and `ESTMP` server always expects UTF-8 encoded messages without advertising `SMTPUTF8`.  This contravenes the RFC's requirements.  Furthermore, the RFC only pertains to ESMTP, as SMTP does not support extensions.

The good news is this is easy to fix: first, instead of decoding as `utf-8`, decode these strings as `self._encoding`.  `SMTP._encoding` should be `ascii`, and `ESMTP._encoding` can be `utf-8`.  Then, add `SMTPUTF8` to [the extensions returned by ESMTP.extensions](https://github.com/twisted/twisted/blob/a8f654d642b6a2a88524dfe690a6a04814752506/src/twisted/mail/smtp.py#L1628-L1639).  This will make both SMTP and ESMTP compliant with my reading of the relevant RFCs.